### PR TITLE
I2c host rx sample

### PR DESCRIPTION
**Pull request by visbaz**
Rx Over Sample Test

### CODE REVIEW

File: hw/ip/i2c/dv/env/seq_lib/i2c_fifo_reset_fmt_vseq.sv [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        if (i == num_wr_bytes) begin
+          `uvm_info(`gfn, $sformatf("\n  transaction WRITE ended %0s", (fmt_item.stop) ?
+              "with STOP, next transaction should begin with START" :
+              "without STOP, next transaction should begin with RSTART"), UVM_DEBUG)
+        end
+        program_format_flag(fmt_item, "program_write_data_to_target");
+      end // for num wr bytes
+      cfg.clk_rst_vif.wait_clks($urandom_range(100, 2000));
+      reset_fmt_fifo();
+      csr_rd_check(.ptr(ral.status.fmtempty), .compare_value(1));
```
Comment (cindychip) [inline hw/ip/i2c/dv/env/seq_lib/i2c_fifo_reset_fmt_vseq.sv:52]: So after FIFO is empty, can we still stream in some data and make sure the previous data are all flushed out of the FIFO?

File: hw/ip/i2c/dv/env/seq_lib/i2c_fifo_reset_fmt_vseq.sv [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    end // for num runs
+    `uvm_info(`gfn, "\n--> end of i2c_fifo_reset_fmt_vseq", UVM_DEBUG)
+
+  endtask : body
+
+  task reset_fmt_fifo();
+    csr_wr(.ptr(ral.fifo_ctrl.fmtrst), .value(1'b1));
+    csr_wr(.ptr(ral.fifo_ctrl.fmtrst), .value(1'b0));
+  endtask : reset_fmt_fifo
```
[the remark this answers is not in the document]
Comment (visbaz) [reply hw/ip/i2c/dv/env/seq_lib/i2c_fifo_reset_fmt_vseq.sv:61]: This task reset_fmt_fifo() is only used in i2c_fifo_reset_fmt_vseq . I shall move it to base seq incase we use it further in any other sequences.
I shall do the same with reset_rx_fifo() to move it to base seq.

[the remark this answers is not in the document]
Comment (cindychip) [reply hw/ip/i2c/dv/env/seq_lib/i2c_fifo_reset_fmt_vseq.sv:61]: Oh I am super sorry, i probably misread the names. So i do not have any preference here ;) If you think it might be used in other task, please feel free to move to the base sequence. Otherwise, it is totally fine for me.

File: hw/ip/i2c/dv/env/seq_lib/i2c_fifo_reset_fmt_vseq.sv (hunk 1 above)
Comment (cindychip) [reply hw/ip/i2c/dv/env/seq_lib/i2c_fifo_reset_fmt_vseq.sv:59]: Actually after reading the spec, I think the FMT FIFO is read out via SDA port.
So if I understand it correctly - if we want to make sure the FMT FIFO data reads out correctly - we will need to eventually read out from the `sda` and `scl` pins ;)
"The I2C reads each format indicator from the head of FMT_FIFO, and processes them in turn. If none of the flags are set for the format indicator, the I2C FSM simply transmits the Format Byte onto the SCL and SDA pins according to the specification, waits for acknowledgement, and then proceeds to the next format indicator. The format flags modulate the behavior as follows."

Comment (visbaz) [reply hw/ip/i2c/dv/env/seq_lib/i2c_fifo_reset_fmt_vseq.sv:59]: as fdata is WO i have added the following checks to the sequence

when random fmt wr bytes are pushed at the end of the loop i check for 
check fmtlvl == num_wr_bytes
check fmtempty == 0 
wait for random clks
check fmtempty == 1

Let me know your thoughts

Comment (cindychip) [reply hw/ip/i2c/dv/env/seq_lib/i2c_fifo_reset_fmt_vseq.sv:59]: Thanks Vis for initializing the google meet conversation!
This is my thought, I think the FMT FIFO is read out via SDA port.
Here is the spec:
"The I2C reads each format indicator from the head of FMT_FIFO, and processes them in turn. If none of the flags are set for the format indicator, the I2C FSM simply transmits the Format Byte onto the SCL and SDA pins according to the specification, waits for acknowledgement, and then proceeds to the next format indicator. The format flags modulate the behavior as follows."

So I was thinking the FIFO can be transferred to SDA and SCL pins eventually. So I think it would be good to drive the I2C sequence and read out the FIFO to make sure the FIFO is cleared when we issue reset.

The testplan mentioned this as well: https://github.com/lowRISC/opentitan/blob/master/hw/ip/i2c/data/i2c_testplan.hjson#L193
I think this sequence shows the entire FMT read/write transaction, but please correct me if I misunderstand anything :)
https://github.com/lowRISC/opentitan/blob/master/hw/ip/i2c/dv/env/seq_lib/i2c_rx_tx_vseq.sv#L76

Comment (cindychip) [reply hw/ip/i2c/dv/env/seq_lib/i2c_fifo_reset_fmt_vseq.sv:59]: Oh I just realize that I am commenting on the sequence regarding this PR: https://github.com/lowRISC/opentitan/pull/11525/files
Do you want to talk in this PR instead? https://github.com/lowRISC/opentitan/pull/11525/files

Comment (visbaz) [reply hw/ip/i2c/dv/env/seq_lib/i2c_fifo_reset_fmt_vseq.sv:59]: > Thanks Vis for initializing the google meet conversation! This is my thought, I think the FMT FIFO is read out via SDA port. Here is the spec: "The I2C reads each format indicator from the head of FMT_FIFO, and processes them in turn. If none of the flags are set for the format indicator, the I2C FSM simply transmits the Format Byte onto the SCL and SDA pins according to the specification, waits for acknowledgement, and then proceeds to the next format indicator. The format flags modulate the behavior as follows."
> 
> So I was thinking the FIFO can be transferred to SDA and SCL pins eventually. So I think it would be good to drive the I2C sequence and read out the FIFO to make sure the FIFO is cleared when we issue reset.
> 
> The testplan mentioned this as well: https://github.com/lowRISC/opentitan/blob/master/hw/ip/i2c/data/i2c_testplan.hjson#L193 I think this sequence shows the entire FMT read/write transaction, but please correct me if I misunderstand anything :) https://github.com/lowRISC/opentitan/blob/master/hw/ip/i2c/dv/env/seq_lib/i2c_rx_tx_vseq.sv#L76

I will check this and fix this way ...

Comment (visbaz) [reply hw/ip/i2c/dv/env/seq_lib/i2c_fifo_reset_fmt_vseq.sv:59]: > Oh I just realize that I am commenting on the sequence regarding this PR: https://github.com/lowRISC/opentitan/pull/11525/files Do you want to talk in this PR instead? https://github.com/lowRISC/opentitan/pull/11525/files

yes we can talk in this PR instead

File: hw/ip/i2c/dv/env/i2c_scoreboard.sv [hunk 3]
```diff
@@ -265,6 +272,10 @@ class i2c_scoreboard extends cip_base_scoreboard #(
           // check in test
           do_read_check = 1'b0;
         end
+        "fifo_status": begin
+          // check in test
+          do_read_check = 1'b0;
+        end
       endcase
```
Comment (jdonjdon) [inline hw/ip/i2c/dv/env/i2c_scoreboard.sv:279]: you probably need default: clause - I know it is dummy - 
to avoid robot crying for lint.

File: hw/ip/i2c/dv/env/seq_lib/i2c_rx_oversample_vseq.sv [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+      t_buf   inside {[(tsu_sta - t_r + 1) :
+                       (tsu_sta - t_r + 1) + cfg.seq_cfg.i2c_time_range]};
+      t_sda_unstable     inside {[0 : t_r + thigh + t_f - 1]};
+      t_sda_interference inside {[0 : t_r + thigh + t_f - 1]};
+      t_scl_interference inside {[0 : t_r + thigh + t_f - 1]};
+    }
+  }
+
+  virtual task body();
+    bit do_interrupt = 1'b1;
```
Comment (jdonjdon) [inline hw/ip/i2c/dv/env/seq_lib/i2c_rx_oversample_vseq.sv:47]: If num_runs > 1,  you can't process_interrupts in your 2nd round due to line 57.
You might want to move this in the loop.

File: hw/ip/i2c/dv/env/i2c_scoreboard.sv (hunk 3 above)
Comment (visbaz) [reply hw/ip/i2c/dv/env/i2c_scoreboard.sv:279]: added default

File: hw/ip/i2c/dv/env/seq_lib/i2c_rx_oversample_vseq.sv (hunk 4 above)
Comment (visbaz) [reply hw/ip/i2c/dv/env/seq_lib/i2c_rx_oversample_vseq.sv:47]: added to the loop

File: hw/ip/i2c/dv/env/seq_lib/i2c_fifo_reset_fmt_vseq.sv (hunk 1 above)
Comment (visbaz) [reply hw/ip/i2c/dv/env/seq_lib/i2c_fifo_reset_fmt_vseq.sv:59]: > > Thanks Vis for initializing the google meet conversation! This is my thought, I think the FMT FIFO is read out via SDA port. Here is the spec: "The I2C reads each format indicator from the head of FMT_FIFO, and processes them in turn. If none of the flags are set for the format indicator, the I2C FSM simply transmits the Format Byte onto the SCL and SDA pins according to the specification, waits for acknowledgement, and then proceeds to the next format indicator. The format flags modulate the behavior as follows."
> > So I was thinking the FIFO can be transferred to SDA and SCL pins eventually. So I think it would be good to drive the I2C sequence and read out the FIFO to make sure the FIFO is cleared when we issue reset.
> > The testplan mentioned this as well: https://github.com/lowRISC/opentitan/blob/master/hw/ip/i2c/data/i2c_testplan.hjson#L193 I think this sequence shows the entire FMT read/write transaction, but please correct me if I misunderstand anything :) https://github.com/lowRISC/opentitan/blob/master/hw/ip/i2c/dv/env/seq_lib/i2c_rx_tx_vseq.sv#L76
> 
> I will check this and fix this way ...

Hi @cindychip FMT fifo gets cleared if we run the sequence of fmt bytes to be sent to the target.
I understand that its important to see that FMT bytes are not processed from the fifo and when FMT reset is applied,
FIFO is cleared.

Basically if you set your test to run WriteOnly trasactions the scenario mentioned by you above will be achieved and when all bytes are transfered to target FMT reset doesnt show an impact as FMT FIFO LVL will be 0 anyways.
i would let the test as it is and if you wish to discuss we can have a quick call.

Comment (visbaz) [reply hw/ip/i2c/dv/env/seq_lib/i2c_fifo_reset_fmt_vseq.sv:59]: @jdonjdon Let me know your thoughts on this discussion

Comment (jdonjdon) [reply hw/ip/i2c/dv/env/seq_lib/i2c_fifo_reset_fmt_vseq.sv:56]: @visbaz As we discussed off line, please verify 'fmtempty' in a designated test.

**Review by jdonjdon — APPROVED** [review]

File: hw/ip/i2c/dv/env/seq_lib/i2c_fifo_reset_fmt_vseq.sv (hunk 1 above)
Comment (visbaz) [reply hw/ip/i2c/dv/env/seq_lib/i2c_fifo_reset_fmt_vseq.sv:56]: @jdonjdon and @cindychip a new test and sequence i2c_fifo_fmt_empty_vseq.sv will be added. I have this ready but please merge this PR and i shall add i2c_fifo_fmt_empty test in new PR.
Will add you both as reviewers for new PR.

**Review by jdonjdon — APPROVED** [review]
